### PR TITLE
Repeat Visitor: Improve validation for numerical input

### PR DESCRIPTION
**Pull request by jsnmoon**
This changes the behavior of the numerical input so that only valid values will be accepted as changes. This prevents a host of issues such as...

- Negative numbers entered into the input. (e.g. `-14`)
- Letters entered into the input. (e.g. `e`)
- Decimal values entered into the input. (e.g. `3.14`)

#### Testing instructions

1. Spin up a [Jurassic Ninja instance](https://jurassic.ninja/create/?gutenpack&calypsobranch=fix/repeat-visitor-input-validation).

2. Connect Jetpack to your WPCOM account.

3. Navigate to Settings->Jetpack Constants, check JETPACK_BETA_BLOCKS, and Save.

4. Add the Repeat Visitor block to a post or page. 

5. Try changing the threshold input value; ensure you can't enter non-numerical/ negative/decimal values.

### CODE REVIEW

**Comment by bluefuton** [comment]
Interestingly, it is still possible to enter a decimal followed by zeros in Chrome, like 2.0 or 5.00:

screen shot 2019-02-27 at 12 57 24

**Comment by jsnmoon** [comment]
> it is still possible to enter a decimal followed by zeros in Chrome, like 2.0 or 5.00 (@bluefuton)

Ah, this is due to the fact that we're checking whether the input's value (cast as a number) is a valid integer. Perhaps it'd be better if we did a regex check on the input string? Like `^\d+$`?

**Comment by bluefuton** [comment]
Nice. That feels better 👍

I found one more edge case: it's possible to pick a zero threshold and 'before' as the criteria:

screen shot 2019-02-27 at 13 45 00

**Comment by jsnmoon** [comment]
> I found one more edge case: it's possible to pick a zero threshold and 'before' as the criteria. (@bluefuton)

Ah, good catch! Fixed.

**Comment by bluefuton** [comment]
Nice! Zero threshold is no longer possible.

One remaining oddity is that it's not possible to delete the current value and enter a new one. For example, if it's currently 3 and I want to enter 7, I can't press backspace/delete to remove 3.

**Comment by jsnmoon** [comment]
> it's not possible to delete the current value and enter a new one (@bluefuton)

This is the consequence borne from these two characteristics of our input validation:

1. The threshold value is directly bound to the numerical input.* 
2. The threshold value validation must be performed while propagating value changes (i.e. numerical and bigger than 0).

(* Due to how `TextControl` is implemented, we're forced to use the `onChange` property to propagate input value changes.)

I went ahead and reverted back to using `defaultValue` and providing a styled `help` message when an invalid value has been entered into the input. 

How's it look?

File: client/gutenberg/extensions/repeat-visitor/components/edit.jsx [hunk 1]
```diff
@@ -72,10 +77,12 @@ class RepeatVisitorEdit extends Component {
 				>
 					<TextControl
 						className="wp-block-jetpack-repeat-visitor-threshold"
-						label={ __( 'Visit count threshold' ) }
 						defaultValue={ this.props.attributes.threshold }
+						help={ this.props.attributes.isThresholdValid ? '' : 'Please enter a valid number.' }
```
Comment (bluefuton) [inline client/gutenberg/extensions/repeat-visitor/components/edit.jsx]:
```suggestion
						help={ this.props.attributes.isThresholdValid ? '' : __( 'Please enter a valid number.' ) }
```

**Pull request merged**